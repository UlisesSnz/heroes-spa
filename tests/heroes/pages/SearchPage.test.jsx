import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <SearchPage />', () => {
    beforeEach(() => jest.clearAllMocks());

    test('debe de mostrarse correctamente con valores por defecto', () => {
    
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        // screen.debug();

        expect(container).toMatchSnapshot();

    });

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {
    
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('/heroes/dc-batman.jpg');

        const alert = screen.getByTestId('alert-danger'); // Con aria-label usamos getByLabelText
        expect(alert.style.display).toBe('none');

    });
    
    test('debe de mostrar un error si no se encuentra el hero (batman123)', () => {
    
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>    
        );

        // screen.debug();

        const alert = screen.getByTestId('alert-danger');
        expect(alert.style.display).not.toBe('none');

    });

    test('debe de llamar el navigate a la pantalla nueva', () => {
        const inputValue = 'batman';

        render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        const btnSearch = screen.getByRole('button');

        fireEvent.change(input, {target: {value: inputValue}});
        fireEvent.click(btnSearch);

        expect(mockedUseNavigate).toHaveBeenCalled();
        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);

    });
        
});
