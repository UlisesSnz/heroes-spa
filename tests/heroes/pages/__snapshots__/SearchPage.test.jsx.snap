// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <div
    class="animate__animated animate__bounceInLeft"
  >
    <h1>
      Search
    </h1>
    <hr />
    <div
      class="row"
    >
      <div
        class="col-5"
      >
        <h4>
          Searching
        </h4>
        <hr />
        <form>
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Search a hero"
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-1"
          >
            Search
          </button>
        </form>
      </div>
      <div
        class="col-7"
      >
        <h4>
          Results
        </h4>
        <hr />
        <div
          class="alert alert-primary animate__animated animate__fadeIn"
        >
          Search a hero
        </div>
        <div
          class="alert alert-danger animate__animated animate__fadeIn"
          data-testid="alert-danger"
          style="display: none;"
        >
          No hero with 
          <b />
        </div>
      </div>
    </div>
  </div>
</div>
`;
